File: src/algorithms/csp/heuristics.js
// src/algorithms/csp/heuristics.js
// Variable and value ordering heuristics for CSP solver.

/**
 * MRV (Minimum Remaining Values) heuristic.
 * Selects the variable (task chunk) with the fewest legal slots remaining.
 * Ties broken by highest priority score.
 *
 * @param {object[]} unassignedVariables - array of { variable, domain }
 * @returns {object | null} the variable object with smallest domain
 */
export function selectVariableMRV(unassignedVariables) {
  if (!unassignedVariables.length) return null;

  let best = null;
  let bestDomainSize = Infinity;
  let bestPriority = -Infinity;

  for (const entry of unassignedVariables) {
    const domainSize = entry.domain.length;
    const priority = entry.variable.priorityScore ?? 0;

    if (
      domainSize < bestDomainSize ||
      (domainSize === bestDomainSize && priority > bestPriority)
    ) {
      best = entry;
      bestDomainSize = domainSize;
      bestPriority = priority;
    }
  }

  return best;
}

/**
 * LCV (Least Constraining Value) heuristic.
 * Orders domain values (slots) by how many options they leave for other variables.
 * In practice, we approximate by preferring slots that:
 *   1. Have lower soft-constraint penalty
 *   2. Start later in the day (leaves earlier slots for urgent tasks)
 *   3. Are on days with more remaining capacity
 *
 * @param {object[]} domain - array of candidate slots
 * @param {function} scoreFn - (slot) => number (lower is better)
 * @returns {object[]} sorted domain (best first)
 */
export function orderValuesLCV(domain, scoreFn) {
  return [...domain].sort((a, b) => {
    const scoreA = scoreFn(a);
    const scoreB = scoreFn(b);
    if (scoreA !== scoreB) return scoreA - scoreB;
    // Tie-break: prefer earlier date (natural order), then earlier time
    // The soft score already penalizes overloaded days, so this is fine
    return a.start.getTime() - b.start.getTime();
  });
}

/**
 * Degree heuristic (tie-breaker for MRV).
 * Prefers variables involved in the most constraints with unassigned variables.
 * For our scheduler, this translates to tasks with:
 *   - Tighter deadlines
 *   - Larger remaining duration (more chunks to place)
 *
 * @param {object} variable
 * @returns {number} degree score (higher = more constrained)
 */
export function computeDegree(variable) {
  let score = 0;

  // Tighter deadline = higher degree
  if (variable.deadline) {
    const daysUntilDeadline = Math.max(
      0,
      (new Date(variable.deadline).getTime() - Date.now()) / (1000 * 60 * 60 * 24)
    );
    score += Math.max(0, 30 - daysUntilDeadline);
  }

  // More remaining chunks = higher degree
  score += (variable.remainingChunks || 1) * 2;

  // Higher priority = higher degree
  score += (variable.priorityScore || 0) / 10;

  return score;
}

/**
 * Forward checking: after assigning a value, prune domains of other variables.
 * Returns the pruned domains or null if any domain becomes empty (failure).
 *
 * @param {object} assignedSlot - the slot just assigned
 * @param {Map<string, object[]>} variableDomains - map of variableId -> domain
 * @param {string} assignedVarId - the variable that was just assigned
 * @returns {Map<string, object[]> | null} pruned domains or null on failure
 */
export function forwardCheck(assignedSlot, variableDomains, assignedVarId) {
  const prunedDomains = new Map();

  for (const [varId, domain] of variableDomains) {
    if (varId === assignedVarId) {
      prunedDomains.set(varId, [assignedSlot]); // assigned
      continue;
    }

    // Remove slots that overlap with the assigned slot
    const filtered = domain.filter((slot) => {
      return slot.end <= assignedSlot.start || slot.start >= assignedSlot.end;
    });

    if (filtered.length === 0) {
      return null; // domain wipeout → backtrack
    }

    prunedDomains.set(varId, filtered);
  }

  return prunedDomains;
}

/**
 * Arc consistency (AC-3) for additional constraint propagation.
 * Reduces domains by ensuring every value has a consistent partner.
 *
 * @param {Map<string, object[]>} variableDomains
 * @returns {Map<string, object[]> | null} reduced domains or null on failure
 */
export function arcConsistency(variableDomains) {
  const queue = [];
  const varIds = [...variableDomains.keys()];

  // Initialize queue with all arcs
  for (let i = 0; i < varIds.length; i++) {
    for (let j = i + 1; j < varIds.length; j++) {
      queue.push([varIds[i], varIds[j]]);
      queue.push([varIds[j], varIds[i]]);
    }
  }

  const domains = new Map(variableDomains);

  while (queue.length > 0) {
    const [xi, xj] = queue.shift();
    if (revise(domains, xi, xj)) {
      if (domains.get(xi).length === 0) {
        return null; // failure
      }
      // Add neighbors back to queue
      for (const xk of varIds) {
        if (xk !== xi && xk !== xj) {
          queue.push([xk, xi]);
        }
      }
    }
  }

  return domains;
}

/**
 * Revise domain of xi to be arc-consistent with xj.
 * @returns {boolean} true if domain was reduced
 */
function revise(domains, xi, xj) {
  let revised = false;
  const domainI = domains.get(xi);
  const domainJ = domains.get(xj);

  const newDomainI = domainI.filter((slotI) => {
    // Check if there's at least one consistent value in xj's domain
    return domainJ.some((slotJ) => {
      return slotI.end <= slotJ.start || slotI.start >= slotJ.end;
    });
  });

  if (newDomainI.length < domainI.length) {
    domains.set(xi, newDomainI);
    revised = true;
  }

  return revised;
}

File: src/algorithms/csp/heuristics.py
# CSP heuristics: MRV, LCV, forward checking, and arc-consistency utilities
from typing import List, Dict, Optional, Tuple
import random
from datetime import datetime


def select_variable_mrv(unassigned_variables: List[Dict], rng: Optional[random.Random] = None) -> Optional[Dict]:
    # Select variable with Minimum Remaining Values (smallest domain)
    if not unassigned_variables:
        return None

    # ORDERING CONSTRAINT: Only consider chunks whose previous chunks are already assigned
    # Filter out chunks that can't be scheduled yet due to ordering
    selectable = []
    for entry in unassigned_variables:
        var = entry.get("variable", {})
        var_id = var.get("id", "")
        
        # Check if this is a chunk variable
        if "_chunk_" in var_id:
            parts = var_id.split("_chunk_")
            if len(parts) == 2:
                task_id = parts[0]
                try:
                    chunk_num = int(parts[1])
                except:
                    chunk_num = 0
                
                # For chunk N, check if all chunks 0..N-1 are already assigned
                # We do this by checking if they're in the unassigned list
                can_select = True
                for prev_chunk in range(chunk_num):
                    prev_chunk_id = f"{task_id}_chunk_{prev_chunk}"
                    # If previous chunk is still unassigned, can't select this one yet
                    if any(e.get("variable", {}).get("id") == prev_chunk_id for e in unassigned_variables):
                        can_select = False
                        break
                
                if can_select:
                    selectable.append(entry)
            else:
                selectable.append(entry)
        else:
            selectable.append(entry)
    
    if not selectable:
        return None

    # CRITICAL FIX: Among selectable variables, prioritize by CHUNK INDEX first
    # This ensures chunk_0 is selected before chunk_1, chunk_1 before chunk_2, etc.
    def get_chunk_index(entry):
        var_id = entry.get("variable", {}).get("id", "")
        if "_chunk_" in var_id:
            parts = var_id.split("_chunk_")
            if len(parts) == 2:
                try:
                    return int(parts[1])
                except:
                    pass
        return 999999  # Non-chunk variables get high index (scheduled last)
    
    # Sort by: chunk index (ascending), then domain size (ascending), then priority (descending)
    selectable_sorted = sorted(
        selectable,
        key=lambda e: (
            get_chunk_index(e),  # Lower chunk indices first
            len(e.get("domain", [])),  # Smaller domain (MRV)
            -e.get("variable", {}).get("priorityScore", 0)  # Higher priority
        )
    )
    
    # Return the first one (lowest chunk index, smallest domain, highest priority)
    return selectable_sorted[0]


def order_values_lcv(domain: List[dict], score_fn, rng: Optional[random.Random] = None) -> List[dict]:
    # Order domain values by Least Constraining Value (LCV) using score_fn
    scored = []
    for s in domain:
        try:
            score = score_fn(s)
        except Exception:
            score = 0
        # small random tie-breaker when rng provided
        tie = rng.random() if rng else 0
        scored.append((s, score, tie))

    scored.sort(key=lambda t: (t[1], t[2], t[0]["start"]))
    return [t[0] for t in scored]


def compute_degree(variable: Dict) -> float:
    # Heuristic degree combining deadline proximity, remaining chunks and priority
    score = 0.0
    if variable.get("deadline"):
        try:
            days_until = max(0.0, (variable["deadline"] - datetime.now()).total_seconds() / (60 * 60 * 24))
        except Exception:
            days_until = 0.0
        score += max(0, 30 - days_until)

    score += (variable.get("remainingChunks", 1) or 1) * 2
    score += (variable.get("priorityScore", 0) or 0) / 10.0
    return score


def forward_check(assigned_slot: dict, variable_domains: Dict[str, List[dict]], assigned_var_id: str) -> Optional[Dict[str, List[dict]]]:
    # Remove conflicting domain values after assigning a slot (forward checking)
    pruned = {}

    for var_id, domain in variable_domains.items():
        if var_id == assigned_var_id:
            pruned[var_id] = [assigned_slot]
            continue

        filtered = [slot for slot in domain if slot["end"] <= assigned_slot["start"] or slot["start"] >= assigned_slot["end"]]

        if not filtered:
            return None

        pruned[var_id] = filtered

    return pruned


def arc_consistency(variable_domains: Dict[str, List[dict]]) -> Optional[Dict[str, List[dict]]]:
    # AC-3 algorithm to enforce arc consistency on variable domains
    from collections import deque

    var_ids = list(variable_domains.keys())
    queue = deque()
    for i in range(len(var_ids)):
        for j in range(len(var_ids)):
            if i == j:
                continue
            queue.append((var_ids[i], var_ids[j]))

    domains = {k: list(v) for k, v in variable_domains.items()}

    def revise(xi, xj):
        revised = False
        new_domain = []
        for slot_i in domains[xi]:
            ok = any(slot_i["end"] <= slot_j["start"] or slot_i["start"] >= slot_j["end"] for slot_j in domains[xj])
            if ok:
                new_domain.append(slot_i)
        if len(new_domain) < len(domains[xi]):
            domains[xi] = new_domain
            revised = True
        return revised

    while queue:
        xi, xj = queue.popleft()
        if revise(xi, xj):
            if not domains[xi]:
                return None
            for xk in var_ids:
                if xk != xi and xk != xj:
                    queue.append((xk, xi))

    return domains
